File: Set Card Game/JavaScript/src/deck.js
import Card from './card.js'

// Deck used to draw cards from during game
export default class Deck {
	constructor() {
		const numbers = [1, 2, 3];
		const symbols = ["diamond", "oval", "squiggle"];
		const colors = ["red", "green", "purple"];
		const shadings = ["open", "lined", "solid"];
		this.new_cards = [];
		// fill new_cards with 81 unique cards (indx corresponds to the name of the .png the card references
		let indx = 0;
		for (let a = 0; a < colors.length; a++) {
			for (let b = 0; b < shadings.length; b++) {
				for (let c = 0; c < symbols.length; c++) {
					for (let d = 0; d < numbers.length; d++) {
						this.new_cards.push(new Card(numbers[d], symbols[c], colors[a], shadings[b], indx));
						indx++;
					}
				}
			}
		}
		this.shuffle();
	}

	// shuffle cards to ensure randomness
	shuffle(){
		let len = this.new_cards.length, temp, i;

		while (len > 0) {
        	i = Math.floor(Math.random() * len);
        	len--;
        	temp = this.new_cards[len];
        	this.new_cards[len] = this.new_cards[i];
        	this.new_cards[i] = temp;
    		}

	}

	// remove cards from deck when dealing
	next_card() {
	    return(this.new_cards.shift());
	}
}

File: Set Card Game/Ruby/deck.rb
# coding: utf-8
# frozen_string_literal: true

require_relative 'card'
require 'colorize'
# Creation of a deck and functions required to manipulate a deck
#
class Deck
  attr_accessor :new_cards
  # A constructor to return a well shuffled deck of 81 cards
  # return: Array of cards
  def initialize
    numbers = [1, 2, 3]
    symbols = %w[⋄ ⌇ ♥]
    colors = %w[red green purple]
    shading = %w[solid striped open]
    @new_cards = []
    numbers.each do |i|
      symbols.each do |j|
        colors.each do |k|
          shading.each do |l|
            @new_cards.push(Card.new(i, j, k, l))
          end
        end
      end
    end
    @new_cards.shuffle!
  end

  # This function returns the next card of the deck
  # return: card 

  def next_card
    @new_cards.shift
  end

  # This function returns the color of the card
  # return: color 

  def colorcard
     @color
  end

  # This function returns true as soon as there is a set in the deck or returns false otherwise
  # return: bool true/false

  def check_if_set
    (0..(@new_cards.length - 1)).each do |i|
      (i + 1..(@new_cards.length - 1)).each do |j|
        (j + 1..(@new_cards.length - 1)).each do |k|
          return true if Card.is_set(@new_cards.at(i), @new_cards.at(j), @new_cards.at(k))
        end
      end
    end
    false
  end
end
